### PR TITLE
Speed up Set index a little

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Speed up indices of type `set` a little (around 1.5 times) by removing garbage.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

<details>
    <summary>Modify your CI run</summary>

**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Unit tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with Analyzer
- [ ]  All with Azure
- [ ]  Add your option here

#### Exclude tests:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with MSAN
- [ ]  All with UBSAN
- [ ]  All with Coverage
- [ ]  All with Aarch64
- [ ]  Add your option here

#### Extra options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

</details>

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
```
CREATE TABLE test (k UInt32, x UInt32, INDEX idx (x) TYPE set(10) GRANULARITY 1) ENGINE = MergeTree ORDER BY k SETTINGS index_granularity = 111;
SET max_insert_threads = 64;
INSERT INTO test SELECT number, number DIV 100 + rand() % 7 FROM numbers(1e9);

SELECT count() FROM test WHERE x = 1234567 SETTINGS send_logs_level = 'trace';
```

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 2a9795f4e39e6b8e2ef0aee3d2e97f396416662e with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64098/2a9795f4e39e6b8e2ef0aee3d2e97f396416662e/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by den-crane** [comment]
BTW, once I noticed that `set` is 4 times slower than `minmax` with the same data.

```sql
set max_threads=1;

create table ind_minmax(A Int64, B Int64, index x1 B type minmax granularity 1) 
Engine MergeTree order by A as select number, cityHash64(number)%111111=1 from numbers(1e8);
optimize table ind_minmax final;

create table ind_set(A Int64, B Int64, index x1 B type set(2) granularity 1) 
Engine MergeTree order by A as select number, cityHash64(number)%111111=1 from numbers(1e8);
optimize table ind_set final;

select count() from ind_minmax where B = 1;
Elapsed: 0.015 sec. Processed 7.26 million rows

select count() from ind_set where B = 1;
Elapsed: 0.060 sec. Processed 7.26 million rows
```

0.015 sec vs 0.060 sec

**Comment by nickitat** [comment]
> BTW, once I noticed that `set` is 4 times slower than `minmax` with the same data.
> 
> ```sql
> set max_threads=1;
> 
> create table ind_minmax(A Int64, B Int64, index x1 B type minmax granularity 1) 
> Engine MergeTree order by A as select number, cityHash64(number)%111111=1 from numbers(1e8);
> optimize table ind_minmax final;
> 
> create table ind_set(A Int64, B Int64, index x1 B type set(2) granularity 1) 
> Engine MergeTree order by A as select number, cityHash64(number)%111111=1 from numbers(1e8);
> optimize table ind_set final;
> 
> select count() from ind_minmax where B = 1;
> Elapsed: 0.015 sec. Processed 7.26 million rows
> 
> select count() from ind_set where B = 1;
> Elapsed: 0.060 sec. Processed 7.26 million rows
> 
> drop table ind_set;
> drop table ind_minmax;
> ```
> 
> 0.015 sec vs 0.060 sec

because we check predicate against all set elements. it is known sub-optimality (lack of min-max for set)

**Comment by alexey-milovidov** [comment]
Only 18%, but still good: https://s3.amazonaws.com/clickhouse-test-reports/64098/dcf410a91673a63612c3915a4fec1e7f0b7ee313/performance_comparison_aarch64_[3_4]/report.html

File: src/Storages/MergeTree/MergeTreeIndexSet.cpp [hunk 1]
```diff
@@ -92,8 +91,6 @@ void MergeTreeIndexGranuleSet::deserializeBinary(ReadBuffer & istr, MergeTreeInd
     if (version != 1)
         throw Exception(ErrorCodes::LOGICAL_ERROR, "Unknown index version {}.", version);
 
-    block.clear();
```
Comment (Algunenano) [inline src/Storages/MergeTree/MergeTreeIndexSet.cpp:95]: This is the cause for https://github.com/ClickHouse/ClickHouse/issues/70108

**Pull request merged**